### ISSUE TITLE
LocalePicker doesn't translate slugs on CMSPage

### ISSUE THREAD

**Issue by CptMeatball** [rainlab/translate-plugin#227]
The JS implementation of the LocalePicker doesn't work flawlessly when using multiple multi language `slugs`.

### Steps to reproduce
Languages (2): NL (default) and DE
CMS URL: `/staticPageName/:category/:slug`
Custom Component that uses both `category` and `slug` parameters.

#### Actions
**Current URL:** `www.domain.tld/de/pagename_de/category_de/slug_de`
-> Switch language through `localePicker` component. 
In this case the URL becomes: `www.domain.tld/nl/pagename_nl/category_de/slug_de`. It fails to set the slugs to the new language.

CMS now tries to query the wrong language in the `transWhere` function of the model. As the default language is selected, it queries for the **untranslated** slugs.

Alternatively, if the non-default language is selected, it queries for the **translated** slugs, but in this case it has a fallback to the default language and the query is successful.

### Expected Behavior
LocalePicker components also translates the `slugs` when switching language.

**Comment by wetgi** [comment]
I can confirm this

**Comment by mariavilaro** [comment]
I'm having the same problem.

**Comment by roulendz** [comment]
the same for me, I am using slug to open translated model, but I get Record not found!
I have custom component with slug and title.
For example
```
/lv/slug //works well
/ru/slug_in_ru // shows Record not found
```

**Comment by mariavilaro** [comment]
I found a workaround for this, but maybe having an indicator that the language has just changed would be better than using the session:

    public function onRun()
    {
        $slug = $this->param('slug');

        if (Session::has('lastLocale') && Session::get('lastLocale') != $this->page['activeLocale']) {
            $news = NewsModel::transWhere('slug', $slug, Session::get('lastLocale'))->first();
            Session::put('lastLocale', $this->page['activeLocale']);
            if ($news) {
                $page = Page::loadCached(Theme::getActiveTheme(), 'news-detail');
                $page->rewriteTranslatablePageUrl();
                $url = Translator::instance()->getPathInLocale(str_replace(':slug', $news->slug, $page->url), $this->page['activeLocale']);
                return Redirect::to($url);
            }
        }

        $news = NewsModel::transWhere('slug', $slug)->first();
        if (!$news) {
            $this->setStatusCode(404);
            return $this->controller->run('404');
        }
        $this->page['news'] = $news;
        Session::put('lastLocale', $this->page['activeLocale']);
    }

**Comment by LukeTowers** [comment]
Thanks for that @mariavilaro! In regards to this issue in general, unfortunately I don't use this plugin on many projects myself so I'm not overly familiar with it or this issue; but if somebody else were to determine the problem and then make a PR detailing how to replicate the problem with the fix as well I'd happily merge it.

**Comment by mariavilaro** [comment]
I tried to do it! I don't know if it's the best way, but it works. The pull request is dirty with other commits I did before because I forgot to did them in another branch, sorry, pick what you need.

**Comment by roulendz** [comment]
@mariavilaro where I should put this code?

**Comment by mariavilaro** [comment]
@roulendz in your component, where you load your data to show on the page, but this is just an example, you need to adapt it to your code/data

**Comment by LukeTowers** [comment]
I'm liking @mariavilaro's solution the most, just #244 to get cleaned up a bit first.

**Comment by quimalguero** [comment]
Hi, we are reading about this issue because we have same problem. I see is not closed yet, this last commit fix the issue?

**Comment by LukeTowers** [comment]
@quimalguero See #341, you can use the implemented event to take care of it.

**Comment by quimalguero** [comment]
Thx @LukeTowers . It worked for me. Gràcies @mariavilaro per la teva aportació.
Now the problem i see is on change language with LocalePicker from that pages, coz localeLinks returns the same url value for all languages (basicaly returns the same value that comes from get). Someone has finded a solution?

**Comment by LukeTowers** [comment]
Not sure what you're talking about @quimalguero

**Comment by davidjmz** [comment]
@LukeTowers I think he means that the locale picker generates always the same url in all languages, so if you change the language inside a post, the url stay the same.

**Comment by LukeTowers** [comment]
Post your code @quimalguero

**Comment by quimalguero** [comment]
@LukeTower

On language selector: 

```
 {% for code, name in locales %}
        <a href="{{''|app}}{{ '/' ~ code ~ localeLinks[code] }}">{{ name }}</a> {% if not loop.last %} | {% endif %}
    {% endfor %} 
```

On page code:

``` 
use nuwadigital\lamasrojana\models\Families;

function onStart() {
      
    
    $slug = $this->param('slug');
    
       if (Session::has('lastLocale') && Session::get('lastLocale') != $this->page['activeLocale']) {
        $fams = Families::transWhere('slug', $slug, Session::get('lastLocale'))->first();
        Session::put('lastLocale', $this->page['activeLocale']);
        if ($fams) {
            $page = Page::loadCached(Theme::getActiveTheme(), 'familia');
            $page->rewriteTranslatablePageUrl();
            $url = Translator::instance()->getPathInLocale(str_replace(':slug', $fams->slug, $page->url), $this->page['activeLocale']);
            return Redirect::to($url);
        }
    }

    $record = Families::transWhere('slug', $slug, Session::get('lastLocale'))->first();
    if (!$record) {
        $this->setStatusCode(404);
        return $this->controller->run('404');
    }
    
    $this->page['Record'] = $record;
    
    Session::put('lastLocale', $this->page['activeLocale']);

}
```

**Comment by LukeTowers** [comment]
Why are you still using that @quimalguero? The merged PR added an event to take care of that.

**Comment by mjauvin** [comment]
This has been fixed by #341

**Comment by MaTToX3** [comment]
Correct me if I am wrong but I have to manually listen to that even for every page that use some sort of slug?
